app_check/src/swig/app_check.i

```diff
@@ -20,14 +20,187 @@
+                            int error_code, const char* error_message) {
+  // Get the function from the map, and erase it
+  auto callback = g_pending_get_tokens[key];
+  g_pending_get_tokens.erase(key);
+
+  AppCheckToken app_check_token;
+  app_check_token.token = token;
+  app_check_token.expire_time_millis = expire_ms;
+  callback(app_check_token, error_code, error_message);
+}
+
+// Wrapper that calls g_get_token_from_csharp. Should be used with the
+// callback logic to guarantee it is on the Unity thread.
+static void CallGetTokenFromCSharp(int key, const char* name) {
+  if (g_get_token_from_csharp) {
+    g_get_token_from_csharp(name, key);
+  } else {
+    // The C# callback has disappeared, so fail the C++ call.
+    FinishGetTokenCallback(key, "", 0, kAppCheckErrorInvalidConfiguration,
+      "Missing AppCheckProvider C# configuration");
+  }
+}
+
+// C++ implementation of the AppCheckProvider that calls up to the
+// C# library.
+class SwigAppCheckProvider : public AppCheckProvider {
+ public:
+  SwigAppCheckProvider(App* app)
+    : app_(app) {}
+
+  ~SwigAppCheckProvider() override {}
+
+  void GetToken(std::function<void(AppCheckToken, int, const std::string&)>
+                  completion_callback) override {
+    if (g_get_token_from_csharp) {
+      // Save the callback in the map, and generate a key
+      int key = g_pending_token_keys++;
+      g_pending_get_tokens[key] = completion_callback;
+      // Queue a call to the C# function that will generate the token.
+      firebase::callback::AddCallback(
+        new firebase::callback::CallbackValue1String1<int>(
+          key, app_->name(), CallGetTokenFromCSharp));
+    } else {
+      completion_callback({}, kAppCheckErrorInvalidConfiguration,
+        "Missing AppCheckProvider C# configuration");
+    }
+  }
+
+ private:
+  App* app_;
+};
+
+// C++ implementation of the AppCheckProviderFactory that can
+// forward to the C# implementation.
+class SwigAppCheckProviderFactory : public AppCheckProviderFactory {
+ public:
+  SwigAppCheckProviderFactory()
+    : provider_map_() {}
+
+  ~SwigAppCheckProviderFactory() override {
+    // Clear the map
+    for (auto it : provider_map_) {
+      delete it.second;
+    }
+    provider_map_.clear();
+  }
+
+  firebase::app_check::AppCheckProvider* CreateProvider(App* app) override {
+    // Check the map
+    std::map<App*, AppCheckProvider*>::iterator it = provider_map_.find(app);
+    if (it != provider_map_.end()) {
+      return it->second;
+    }
+    // Create a new provider and cache it
+    SwigAppCheckProvider* provider = new SwigAppCheckProvider(app);
+    provider_map_[app] = provider;
+    return provider;
+  }
+ private:
+  std::map<App*, AppCheckProvider*> provider_map_;
+};
+
+static SwigAppCheckProviderFactory g_swig_factory;
+
+// Called from C# to register the C# to call to get a token.
+void SetGetTokenCallback(GetTokenFromCSharp get_token_callback) {
```

Comment:
Why do we need a method to set g_get_token_from_csharp?
Can it not always be equal to getTokenDelegate?
In fact, can we remove one of g_get_token_from_csharp or getTokenDelegate?
(It would be fine to still set the native factory to g_swig_factory or to null, or to conditionally do nothing if the C# factory is null)

Basically, because these similar but subtly different variables exist, it is harder for me to follow which method is being called, and hard for me to say when they are equal or different.

Reply:
Or, even more direct, is it possible to call GetTokenFromCSharpMethod by name instead of having a variable like g_get_token_from_csharp?

It seems like we always want GetToken to call GetTokenFromCSharpMethod, so I wonder what is the point of the variable that decides which C# method to call.

Reply:
You've hit on the problem with your second comment.  As far as I know, there isn't a way for C++ to just be aware of the existence of the C# functions.  So instead we have C# call into C++, passing in function pointers to the C# functions, for C++ to then call into.